File: agent/src/main/java/pt/um/lei/masb/agent/data/transaction/ontology/Transaction.java
package pt.um.lei.masb.agent.data.transaction.ontology;


import jade.content.Concept;
import pt.um.lei.masb.blockchain.data.SensorData;

import java.util.Arrays;
import java.util.Objects;

/**
 * Transaction in ontology bean form.
 */
public class Transaction implements Concept {
    private String transactionId;
    private String publicKey;
    private SensorData sd;
    private byte[] signature;

    public Transaction(String transactionId,
                       String publicKey,
                       SensorData sd,
                       byte[] signature) {
        this.transactionId = transactionId;
        this.publicKey = publicKey;
        this.sd = sd;
        this.signature = signature;
    }

    public String getTransactionId() {
        return transactionId;
    }

    public void setTransactionId(String transactionId) {
        this.transactionId = transactionId;
    }

    public String getPublicKey() {
        return publicKey;
    }

    public void setPublicKey(String publicKey) {
        this.publicKey = publicKey;
    }

    public SensorData getSd() {
        return sd;
    }

    public void setSd(SensorData sd) {
        this.sd = sd;
    }

    public byte[] getSignature() {
        return signature;
    }

    public void setSignature(byte[] signature) {
        this.signature = signature;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Transaction that = (Transaction) o;
        return Objects.equals(transactionId, that.transactionId) &&
                Objects.equals(publicKey, that.publicKey) &&
                Objects.equals(sd, that.sd) &&
                Arrays.equals(signature, that.signature);
    }

    @Override
    public int hashCode() {

        int result = Objects.hash(transactionId, publicKey, sd);
        result = 31 * result + Arrays.hashCode(signature);
        return result;
    }

    @Override
    public String toString() {
        return "Transaction{" +
                "transactionId='" + transactionId + '\'' +
                ", publicKey='" + publicKey + '\'' +
                ", sd=" + sd +
                ", signature=" + signature +
                '}';
    }
}

File: ledger/src/main/kotlin/pt/um/lei/masb/blockchain/ledger/Transaction.kt
package pt.um.lei.masb.blockchain.ledger

import com.orientechnologies.orient.core.record.OElement
import mu.KLogging
import org.openjdk.jol.info.ClassLayout
import pt.um.lei.masb.blockchain.data.PhysicalData
import pt.um.lei.masb.blockchain.persistance.NewInstanceSession
import pt.um.lei.masb.blockchain.persistance.Storable
import pt.um.lei.masb.blockchain.service.Ident
import pt.um.lei.masb.blockchain.utils.Crypter
import pt.um.lei.masb.blockchain.utils.DEFAULT_CRYPTER
import pt.um.lei.masb.blockchain.utils.Hashable
import pt.um.lei.masb.blockchain.utils.asHex
import pt.um.lei.masb.blockchain.utils.flattenBytes
import pt.um.lei.masb.blockchain.utils.generateSignature
import pt.um.lei.masb.blockchain.utils.verifyECDSASig
import java.security.PrivateKey
import java.security.PublicKey

data class Transaction(
    // Agent's pub key.
    val publicKey: PublicKey,
    val data: PhysicalData,
    // This is to identify unequivocally an agent.
    val signature: ByteArray
) : Sizeable,
    Hashed,
    Hashable,
    Storable,
    LedgerContract {


    // This is also the hash of the transaction.
    override val hashId = digest(crypter)

    /**
     * Calculate the approximate size of the transaction.
     *
     * @return The size of the transaction in bytes.
     */
    override val approximateSize: Long
        get() = ClassLayout
            .parseClass(this::class.java)
            .instanceSize() + data.approximateSize


    constructor(
        ident: Ident,
        data: PhysicalData
    ) : this(
        ident.publicKey,
        data,
        generateSignature(
            ident.privateKey,
            ident.publicKey,
            data
        )
    )

    constructor(
        privateKey: PrivateKey,
        publicKey: PublicKey,
        data: PhysicalData
    ) : this(
        publicKey,
        data,
        generateSignature(
            privateKey,
            publicKey,
            data
        )
    )

    override fun store(
        session: NewInstanceSession
    ): OElement =
        session
            .newInstance("Transaction")
            .apply {
                setProperty("publicKey", publicKey.encoded)
                setProperty("data", data.store(session))
                setProperty(
                    "signature",
                    session.newInstance(
                        signature
                    )
                )
                setProperty("hashId", hashId)
            }

    /**
     * Verifies the data we signed hasn't been
     * tampered with.
     *
     * @return Whether the data was signed with the
     * corresponding private key.
     */
    fun verifySignature(): Boolean =
        verifyECDSASig(
            publicKey,
            publicKey.encoded + data.digest(crypter),
            signature
        )

    /**
     * TODO: Transaction verification.
     * @return Whether the transaction is valid.
     */
    fun processTransaction(): Boolean {
        return verifySignature()
    }


    override fun toString(): String = """
            |       Transaction {
            |           Transaction id: ${hashId.print()},
            |           Public Key: ${publicKey.encoded.asHex()},
            |$data,
            |           Signature: ${signature.asHex()}
            |       }
            """.trimMargin()

    override fun digest(c: Crypter): Hash =
        c.applyHash(
            flattenBytes(
                publicKey.encoded,
                data.digest(c),
                signature
            )
        )


    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Transaction) return false

        if (!publicKey.encoded!!.contentEquals(
                other.publicKey.encoded
            )
        ) return false
        if (data != other.data) return false
        if (!signature.contentEquals(
                other.signature
            )
        ) return false
        if (!hashId.contentEquals(
                other.hashId
            )
        ) return false

        return true
    }

    override fun hashCode(): Int {
        var result = publicKey.hashCode()
        result = 31 * result + data.hashCode()
        result = 31 * result + signature.contentHashCode()
        result = 31 * result + hashId.contentHashCode()
        return result
    }

    companion object : KLogging() {
        val crypter = DEFAULT_CRYPTER
    }

}
